src/main/java/io/pravega/connectors/flink/FlinkPravegaWriter.java

```diff
@@ -531,33 +539,103 @@ public void restoreState(List<UUID> list) throws Exception {
+                return;
+            }
+
+            Exception exception = null;
+
+            EventStreamWriter<T> restorePravegaWriter = null;
+
+            ClientFactory restoreClientFactory = null;
+
+            Map<String, Map<String, List<PendingTransaction>>> pendingTransactionsMap =
+                    pendingTransactionList.stream().collect(groupingBy(PendingTransaction::getScope, groupingBy(PendingTransaction::getStream)));
+
+            log.debug("pendingTransactionsMap:: " + pendingTransactionsMap);
+
+            for (Map.Entry<String, Map<String, List<PendingTransaction>>> pendingTransactionEntry: pendingTransactionsMap.entrySet()) {
 
-                        if (status == Transaction.Status.OPEN) {
-                            // that is the case when a crash happened between when the master committed
-                            // the checkpoint, and the sink could be notified
-                            log.info("{} - committing completed checkpoint transaction {} after task restore",
-                                    name(), txnId);
+                String scope = pendingTransactionEntry.getKey();
 
-                            txn.commit();
+                Map<String, List<PendingTransaction>> transactionsMap = pendingTransactionEntry.getValue();
 
-                            log.debug("{} - committed checkpoint transaction {}", name(), txnId);
+                for (Map.Entry<String, List<PendingTransaction>> transactionsEntry: transactionsMap.entrySet()) {
 
-                        } else if (status == Transaction.Status.COMMITTED || status == Transaction.Status.COMMITTING) {
-                            // that the common case
-                            log.debug("{} - at restore, transaction {} was already committed", name(), txnId);
+                    String stream = transactionsEntry.getKey();
+
+                    try {
+                        Serializer<T> eventSerializer = new FlinkSerializer<>(serializationSchema);
+                        EventWriterConfig writerConfig = EventWriterConfig.builder()
+                                .transactionTimeoutTime(txnTimeoutMillis)
+                                .transactionTimeoutScaleGracePeriod(txnGracePeriodMillis)
+                                .build();
+                        restoreClientFactory = createClientFactory(scope, controllerURI);
+                        restorePravegaWriter = restoreClientFactory.createEventWriter(stream,
+                                eventSerializer,
+                                writerConfig);
+                        log.info("restore state for the scope: {} and stream: {}", scope, stream);
+
+                        List<PendingTransaction> pendingTransactions = transactionsEntry.getValue();
+
+                        for (PendingTransaction pendingTransaction : pendingTransactions) {
+                            UUID txnId = pendingTransaction.getUuid();
+                            if (txnId != null) {
```

Comment:
Please don't check for null here, instead add a precondition to the `PendingTransaction` constructor that validates the `txnId`.